pilot/pkg/model/sidecar.go

```diff
@@ -0,0 +1,324 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"strings"
+
+	networking "istio.io/api/networking/v1alpha3"
+)
+
+const (
+	wildcardNamespace = "*"
+	wildcardService   = Hostname("*")
+)
+
+// SidecarScope is a wrapper over the Sidecar resource with some
+// preprocessed data to determine the list of services, virtualServices,
+// and destinationRules that are accessible to a given sidecar.
+type SidecarScope struct {
+	// The crd itself. Can be nil if we are constructing the default
+	// sidecar scope
+	Config *Config
+
+	// set of egress listeners, and their associated services.
+	// A sidecar scope should have either ingress/egress listeners or both.
+	EgressListeners []*IstioEgressListenerWrapper
+
+	// // set of ingress listeners
+	// // A sidecar scope should have either ingress/egress listeners or both.
+	// IngressListeners []*IstioIngressListenerWrapper
+
+	// Union of services imported across all egress listeners for use by CDS code.
+	// Right now, we include all the ports in these services.
+	// TODO: Trim the ports in the services to only those referred to by the
+	// egress listeners.
+	allImportedServices []*Service
+
+	// Destination rules imported across all egress listeners
+	allImportedDestinationRules map[Hostname]*Config
+
+	// VirtualServices imported across all egress listeners
+	allImportedVirtualServices []Config
+}
+
+// IstioEgressListenerWrapper is a wrapper for networking.IstioEgressListener object.
+// It has the parsed form of the hosts field, encompassing a list of services
+// per sidecar listener object.
+type IstioEgressListenerWrapper struct {
+	// The actual IstioEgressListener api object from the Config. It can be nil if
+	// this is for the default sidecar scope.
+	IstioListener *networking.IstioEgressListener
```

Comment:
Is this necessary? Since below `ListenerPort ` and `importMap ` exists?

Reply:
Yeah because we have unix domain sockets and stuff in the listener. This is temporary, until I parse out all the fields in the listener into individual fields in the Listener wrapper.

Reply:
I'm also a bit uncomfortable with the wrappers - including the existing ones, that obscure and
translate the resources to a model that no longer matches well with the original config. I think the wrappers are a large part of what makes this PR so hard to parse.

Reply:
There is no ListenerPort anymore. Import map is just for performance so that we dont have to parse strings again and again